pkg/controllers/ippool/ippool_controller.go

```diff
@@ -0,0 +1,249 @@
+	Service *ippool.IPPoolService
+}
+
+func deleteSuccess(r *Reconciler, _ *context.Context, _ *v1alpha2.IPPool) {
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, MetricResType)
+}
+
+func deleteFail(r *Reconciler, c *context.Context, o *v1alpha2.IPPool, e *error) {
+	r.setReadyStatusFalse(c, o, e)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, MetricResType)
+}
+
+func updateSuccess(r *Reconciler, c *context.Context, o *v1alpha2.IPPool) {
+	r.setReadyStatusTrue(c, o)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateSuccessTotal, MetricResType)
+}
+
+func updateFail(r *Reconciler, c *context.Context, o *v1alpha2.IPPool, e *error) {
+	r.setReadyStatusFalse(c, o, e)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateFailTotal, MetricResType)
+}
+
+func (r *Reconciler) setReadyStatusFalse(ctx *context.Context, ippool *v1alpha2.IPPool, err *error) {
+	conditions := []v1alpha1.Condition{
+		{
+			Type:    v1alpha1.Ready,
+			Status:  v1.ConditionFalse,
+			Message: "NSX IPPool could not be created/updated",
+			Reason: fmt.Sprintf(
+				"error occurred while processing the IPPool CR. Error: %v",
+				*err,
+			),
+		},
+	}
+	ippool.Status.Conditions = conditions
+	if ippool.Status.Subnets == nil {
+		ippool.Status.Subnets = make([]v1alpha2.SubnetResult, 0)
+	}
+	e := r.Client.Status().Update(*ctx, ippool)
+	if e != nil {
+		log.Error(e, "unable to update IPPool status", "ippool", ippool)
+	}
+}
+
+func (r *Reconciler) setReadyStatusTrue(ctx *context.Context, ippool *v1alpha2.IPPool) {
+	conditions := []v1alpha1.Condition{
+		{
+			Type:    v1alpha1.Ready,
+			Status:  v1.ConditionTrue,
+			Message: "NSX IPPool has been successfully created/updated",
+			Reason:  "NSX API returned 200 response code for PATCH",
+		},
+	}
+	ippool.Status.Conditions = conditions
+	e := r.Client.Status().Update(*ctx, ippool)
+	if e != nil {
+		log.Error(e, "unable to update IPPool status", "ippool", ippool)
+	}
+}
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	obj := &v1alpha2.IPPool{}
+	log.Info("reconciling ippool CR", "ippool", req.NamespacedName)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, MetricResType)
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch ippool CR", "req", req.NamespacedName)
+		return resultNormal, client.IgnoreNotFound(err)
+	}
+
+	// TODO: Since IPPool v1alpha2 service can only be activated from NSX 4.1.0 onwards,
+	// So need to check NSX version before starting IPPool reconcile
+	//if !r.Service.NSXClient.NSXCheckVersionForStaticRoute() {
+	//	err := errors.New("NSX version check failed, IPPool v1alpha2 is not supported")
+	//	updateFail(r, &ctx, obj, &err)
+	//	if NSX version check fails, it will be put back to reconcile queue and be reconciled after 5 minutes
+	//return ResultRequeueAfter5mins, nil
```

Comment:
why we remove this?

Reply:
It is TODO, we should reorganize NSXCheckVersion, otherwise, every CRD needs to create duplicated logic.